proposals/webidl-bindings/idl.md

```diff
@@ -0,0 +1,739 @@
+>context, all functions simply return a tuple of values. However, in order to
+>support interoperability with systems that are not inherently stack machines,
+>this specification assumes all functions return a single value -- which may be
+>a tuple. In particular, so-called void functions are assumed to return the
+>empty tuple.
+
+A function that may raise an exception can declare it with a throws clause in
+the type:
+
+```
+FunctionType => TupleType '=>' Type 'throws' Type
+```
+or in W-Expression form:
+
+```
+W-FunctionType ::= '(' 'func' W-TupleType W-Type W-Type ')'
+```
+
+The type of an exception is not distinguished from other types. I.e., there is
+no privileged type that is used for exceptions.
+
+### Quantified Type {#quantified-type}
+
+There are two forms of quantifier: universal and existential. 
+
+```
+UniversalType => 'all' TypeVar .. TypeVar '.' Type
+```
+
+and
+
+```
+ExistentialType => 'exist' TypeVar .. TypeVar '.' Type
+```
+
+Or, as W-Expressions:
+
+```
+W-UniversalType ::= '(' 'all' '(' W-TypeVar .. W-TypeVar ')' W-Type ')'
+W-ExistentialType ::= '(' 'exist' '(' W-TypeVar .. W-TypeVar ')' W-Type ')'
+```
+
+Where more than one quantifier is applied to a type, they are left-concatenated; for example:
+
+```
+all X Y . (X Y)
+```
+
+is equivalent to
+
+```
+all X . all Y . (X Y)
+```
+
+There are two forms of TypeVar, corresponding to whether the type variable
+denotes a type or a type function:
+
+```
+TypeVar => Name
+TypeVar => Name '/' Integer
+```
+
+```
+W-TypeVar ::= W-Identifier
+W-TypeVar ::= '(' W-Identifier '/' W-Integer ')'
+```
+
+The second form is used to denote a generic type; for example `Sequence/1`
+denotes the sequence type where the `/1` denotes the fact that `Sequence` takes
+a single type argument.
+
+>Question: Should type hiding be supported? I.e., if a quantified form introduces
+>a type variable that is already in scope does it have the effect of occluding
+>the outer name. Convenience and composability says yes, security says no.
+
+#### Interface {#interface}
+
+An interface type is a collection of type signatures, as an unordered set:
+
+```
+TypeInterface => '{' Signature ..  Signature '}'
+```
+
+```
+W-TypeInterface ::= '{' W-Signature .. W-Signature '}'
+```
+
+There are two forms of Signature; one denotes the type of a field: i.e., it’s an
+association of a type to a name:
+
+```
+Signature => Name ':' Type
+```
+
+```
+W-Signature ::= '(' W-Identifier W-Type ')'
+```
+
+The second form of TypeSignature denotes a nominal type, either implicitly:
+
+```
+Signature => 'type' NominalType
+```
+
+```
+W-Signature ::= '(' 'type' W-NominalType ')'
+```
+
+This form is used to identify so-called type imports. It is also used to
+identify nominal types; in which case it may be associated with an algebraic
+type combination:
+
+```
+Signature => 'type' NominalType '=' Type
+Signature => 'all' TypeVar .. TypeVar '.' NominalType '=' Type
```

Comment:
If you allow type parameters without restricting type recursion then type equality is undecidable.

Also, the `all` is odd here, since this is a lambda, not a quantifier.